File: packages/core/src/format/format.test.ts
import { describe, expect, it } from 'vitest';
import { parseDotenv } from '../importer/dotenv';
import { parseJson } from '../importer/json';
import { parseShell } from '../importer/shell';
import { parseYaml } from '../importer/yaml';
import type { ParseResult } from '../importer/types';
import { ExportFormatError, formatSecrets } from './format';
import type { ExportableSecret } from './format';

const PEM = `-----BEGIN RSA PRIVATE KEY-----
MIIEowIBAAKCAQEAyLGqLTPGWvKmYQF9vB3nJZ0X1kQ7hRr2mN8sD4tW6uV0pC1e
9aXbYzR3kFqL5oN7wS2dH8jT4gM1vB6cP0rE3nK9xA7yU2iO5lQ8fD1zJ4hG6mV3
-----END RSA PRIVATE KEY-----`;

/**
 * Every value that has ever broken a naive quoting implementation. The
 * round-trip assertions below are the real specification of the escaping rules
 * in `format.ts`: a bug there does not throw, it hands someone a password with
 * a character missing.
 */
const NASTY: ExportableSecret[] = [
  { name: 'EMPTY', value: '' },
  { name: 'PLAIN', value: 'simple-value' },
  { name: 'SPACES', value: '  leading and trailing  ' },
  { name: 'HASH', value: 'SG.abc#123' },
  { name: 'HASH_AFTER_SPACE', value: 'value # not a comment' },
  { name: 'EQUALS', value: 'a=b=c' },
  { name: 'SINGLE_QUOTE', value: "it's" },
  { name: 'DOUBLE_QUOTE', value: 'say "hi"' },
  { name: 'BOTH_QUOTES', value: `mixed '"' quotes` },
  { name: 'BACKSLASH', value: 'C:\\Users\\deploy\\.ssh' },
  { name: 'ESCAPE_LOOKALIKE', value: String.raw`literal \n and \t` },
  { name: 'DOLLAR', value: '${NOT_INTERPOLATED} $PATH' },
  { name: 'BACKTICK', value: 'a`whoami`b' },
  { name: 'NEWLINES', value: 'line one\nline two' },
  { name: 'CARRIAGE_RETURN', value: 'before\rafter' },
  { name: 'TAB', value: 'before\tafter' },
  { name: 'UNICODE', value: 'pässwörd — 日本語 — 🔐' },
  { name: 'ONLY_HASH', value: '#' },
  { name: 'PEM_KEY', value: PEM },
];

function asRecord(secrets: ReadonlyArray<ExportableSecret>): Record<string, string> {
  return Object.fromEntries(secrets.map((secret) => [secret.name, secret.value]));
}

function parsedRecord(result: ParseResult): Record<string, string> {
  return Object.fromEntries(result.entries.map((entry) => [entry.key, entry.value]));
}

/**
 * Values a format is honest about being unable to carry, and refuses. They are
 * excluded from that format's round-trip and asserted on separately, so the
 * exclusion is a documented limit rather than a gap in the test.
 */
const UNREPRESENTABLE: Partial<Record<string, readonly string[]>> = {
  shell: ['CARRIAGE_RETURN'],
};

describe('formatSecrets round-trips', () => {
  it.each([
    ['env', parseDotenv],
    ['shell', parseShell],
    ['json', parseJson],
    ['yaml', parseYaml],
  ] as const)('every value survives %s exactly', (format, parse) => {
    const excluded = UNREPRESENTABLE[format] ?? [];
    const secrets = NASTY.filter((secret) => !excluded.includes(secret.name));
    const result = parse(formatSecrets(secrets, format));

    expect(result.warnings).toEqual([]);
    expect(parsedRecord(result)).toEqual(asRecord(secrets));
  });

  it('keeps line-oriented formats in the caller order', () => {
    const result = parseDotenv(formatSecrets(NASTY, 'env'));
    expect(result.entries.map((entry) => entry.key)).toEqual(NASTY.map((secret) => secret.name));
  });

  it('round-trips a value through env, back out, and in again unchanged', () => {
    const once = formatSecrets(NASTY, 'env');
    const twice = formatSecrets(
      parseDotenv(once).entries.map((entry) => ({ name: entry.key, value: entry.value })),
      'env',
    );

    expect(twice).toBe(once);
  });
});

describe('formatSecrets: env', () => {
  it('quotes only when the value requires it', () => {
    const output = formatSecrets(
      [
        { name: 'PLAIN', value: 'simple-value' },
        { name: 'URL', value: 'postgres://app:pw@db.internal:5432/app?sslmode=require' },
        { name: 'EMPTY', value: '' },
        { name: 'SPACED', value: 'two words' },
      ],
      'env',
    );

    expect(output).toBe(
      [
        'PLAIN=simple-value',
        'URL=postgres://app:pw@db.internal:5432/app?sslmode=require',
        'EMPTY=',
        "SPACED='two words'",
        '',
      ].join('\n'),
    );
  });

  // Single quotes are literal in every dialect, so a value containing `$` cannot
  // be expanded by a reader that performs interpolation.
  it('prefers single quotes, which no reader interpolates', () => {
    expect(formatSecrets([{ name: 'A', value: '${HOME}/x' }], 'env')).toBe("A='${HOME}/x'\n");
    expect(formatSecrets([{ name: 'A', value: 'C:\\Users' }], 'env')).toBe("A='C:\\Users'\n");
  });

  // A single-quoted string has no way to escape its own delimiter.
  it('uses double quotes for a value containing an apostrophe', () => {
    expect(formatSecrets([{ name: 'A', value: "it's" }], 'env')).toBe('A="it\'s"\n');
  });

  // One line of output stays one secret: a file truncated mid-write loses a
  // secret instead of merging the rest into the previous value.
  it('escapes newlines rather than writing them literally', () => {
    expect(formatSecrets([{ name: 'A', value: 'one\ntwo' }], 'env')).toBe('A="one\\ntwo"\n');
    expect(formatSecrets([{ name: 'K', value: PEM }], 'env').split('\n')).toHaveLength(2);
  });
});

describe('formatSecrets: shell', () => {
  it('single-quotes every value and escapes embedded quotes', () => {
    const output = formatSecrets(
      [
        { name: 'PLAIN', value: 'value' },
        { name: 'APOSTROPHE', value: "it's" },
        { name: 'INJECTION', value: '$(rm -rf /)' },
      ],
      'shell',
    );

    expect(output).toBe(
      [
        "export PLAIN='value'",
        "export APOSTROPHE='it'\\''s'",
        "export INJECTION='$(rm -rf /)'",
        '',
      ].join('\n'),
    );
  });

  it('writes a newline literally, since single quotes need no escape for it', () => {
    const output = formatSecrets([{ name: 'K', value: PEM }], 'shell');
    expect(parsedRecord(parseShell(output))).toEqual({ K: PEM });
  });

  // Single quotes have no escape mechanism, so the CR would be written raw and
  // normalised away by every reader — including this repository's parsers.
  it('refuses a carriage return rather than losing it on the way back in', () => {
    expect(() => formatSecrets([{ name: 'A', value: 'before\rafter' }], 'shell')).toThrow(
      /carriage return/,
    );
    expect(parsedRecord(parseDotenv(formatSecrets([{ name: 'A', value: 'a\rb' }], 'env')))).toEqual(
      {
        A: 'a\rb',
      },
    );
  });
});

describe('formatSecrets: json', () => {
  it('sorts keys so a diff shows only what changed', () => {
    const output = formatSecrets(
      [
        { name: 'ZULU', value: '1' },
        { name: 'ALPHA', value: '2' },
        { name: 'MIKE', value: '3' },
      ],
      'json',
    );

    expect(output).toBe('{\n  "ALPHA": "2",\n  "MIKE": "3",\n  "ZULU": "1"\n}\n');
  });

  it('produces identical output regardless of the order it was given', () => {
    const forwards = formatSecrets(NASTY, 'json');
    const backwards = formatSecrets([...NASTY].reverse(), 'json');

    expect(forwards).toBe(backwards);
  });
});

describe('formatSecrets: yaml', () => {
  // Unquoted, `yes` is a string under YAML 1.2 and boolean true under 1.1,
  // which is what PyYAML implements.
  it('quotes every value so no reader can retype it', () => {
    const output = formatSecrets(
      [
        { name: 'FLAG', value: 'yes' },
        { name: 'PORT', value: '5432' },
        { name: 'TIME', value: '08:00' },
        { name: 'NOTHING', value: '' },
      ],
      'yaml',
    );

    expect(output).toBe('FLAG: "yes"\nNOTHING: ""\nPORT: "5432"\nTIME: "08:00"\n');
  });

  it('keeps a long value on one line rather than folding it', () => {
    const value = `${'a'.repeat(200)} ${'b'.repeat(200)}`;
    const output = formatSecrets([{ name: 'LONG', value }], 'yaml');

    expect(output.split('\n')).toHaveLength(2);
    expect(parsedRecord(parseYaml(output))).toEqual({ LONG: value });
  });
});

describe('formatSecrets: docker', () => {
  it('writes values verbatim, as --env-file reads them', () => {
    const output = formatSecrets(
      [
        { name: 'PLAIN', value: 'value' },
        { name: 'SPACED', value: 'two words' },
        { name: 'QUOTES', value: `"quoted"` },
      ],
      'docker',
    );

    expect(output).toBe('PLAIN=value\nSPACED=two words\nQUOTES="quoted"\n');
  });

  // Docker would read the remainder of a multi-line value as a new variable, or
  // drop it. Emitting a secret silently cut at the first newline is the failure
  // this module exists to prevent.
  it('refuses a value it cannot represent instead of truncating it', () => {
    expect(() => formatSecrets([{ name: 'PEM_KEY', value: PEM }], 'docker')).toThrow(
      ExportFormatError,
    );
    expect(() => formatSecrets([{ name: 'A', value: 'a\rb' }], 'docker')).toThrow(/line break/);
  });

  it('names the secret but never its value in the error', () => {
    try {
      formatSecrets([{ name: 'STRIPE_KEY', value: 'sk_live_secret\nmore' }], 'docker');
      expect.unreachable('formatSecrets should have thrown');
    } catch (error) {
      expect(error).toBeInstanceOf(ExportFormatError);
      expect((error as ExportFormatError).secretName).toBe('STRIPE_KEY');
      expect((error as Error).message).toContain('STRIPE_KEY');
      expect((error as Error).message).not.toContain('sk_live_secret');
    }
  });
});

describe('formatSecrets', () => {
  it.each(['env', 'json', 'yaml', 'shell', 'docker'] as const)(
    'renders an empty list as an empty %s document',
    (format) => {
      const output = formatSecrets([], format);
      expect(output).toBe(format === 'json' ? '{}\n' : format === 'yaml' ? '{}\n' : '');
    },
  );

  // `A=B=value` would parse back as `A` holding `B=value`.
  it.each(['env', 'json', 'yaml', 'shell', 'docker'] as const)(
    'refuses a name that would not survive %s output',
    (format) => {
      expect(() => formatSecrets([{ name: 'A=B', value: 'x' }], format)).toThrow(ExportFormatError);
      expect(() => formatSecrets([{ name: '', value: 'x' }], format)).toThrow(ExportFormatError);
    },
  );
});

File: cli/cmd/xecret/format_test.go
package main

import (
	"strings"
	"testing"

	"github.com/playxoft/xecret/cli/internal/importer"
)

// The one property `format.go` exists to hold: **whatever goes in comes back
// out, byte for byte**.
//
// A quoting bug here does not throw. It hands somebody a password with a
// character missing, and they find out when a service fails to authenticate at
// 3am. So the test is not a set of expected strings — those only check that the
// output looks like it did yesterday — it is a round trip through this
// repository's own parsers, over values chosen to be hostile.
//
// This is the same test the TypeScript side runs over the same values, and the
// parsers on both ends of the round trip are themselves pinned to each other by
// `internal/importer/fixtures_test.go`. Rendering here and parsing there is what
// makes `xecret pull --format env` on an e2ee environment produce a file the
// dashboard's importer reads identically.

// hostile is every value that has ever broken a `.env` file.
var hostile = map[string]string{
	"SIMPLE":          "hunter2",
	"EMPTY":           "",
	"SPACES":          "  padded value  ",
	"HASH":            "abc#123",
	"HASH_SPACED":     "value # not a comment",
	"SINGLE_QUOTE":    "it's fine",
	"DOUBLE_QUOTE":    `say "hi"`,
	"BOTH_QUOTES":     `it's "both"`,
	"BACKSLASH":       `C:\Users\deploy`,
	"REGEX":           `^\d{3}-\d{4}$`,
	"DOLLAR":          "cost $5 and ${NOT_INTERPOLATED}",
	"BACKTICK":        "`whoami`",
	"NEWLINE":         "-----BEGIN KEY-----\nline one\nline two\n-----END KEY-----",
	"TAB":             "before\tafter",
	"NON_ASCII":       "café ☕ 秘密 naïve",
	"EQUALS":          "a=b=c",
	"LEADING_HASH":    "#starts-with-hash",
	"ONLY_WHITESPACE": "   ",
	"URL":             "postgres://app:p@ss w0rd!@db.example.com:5432/app?sslmode=require",
}

// carriageReturn is separated because two formats refuse it, deliberately.
const carriageReturn = "before\rafter"

func TestEveryFormatRoundTripsHostileValues(t *testing.T) {
	for _, target := range []struct {
		format string
		parse  func(string) importer.Result
	}{
		{"env", func(s string) importer.Result { return importer.ParseDotenv(s) }},
		{"shell", func(s string) importer.Result { return importer.ParseShell(s) }},
		{"json", func(s string) importer.Result { return importer.ParseJSON(s) }},
		{"yaml", func(s string) importer.Result { return importer.ParseYAML(s) }},
	} {
		document, err := formatSecrets(hostile, target.format)
		if err != nil {
			t.Fatalf("%s: %v", target.format, err)
		}

		parsed := target.parse(string(document))
		if len(parsed.Warnings) > 0 {
			t.Errorf("%s: round trip produced warnings: %v", target.format, parsed.Warnings)
		}

		got := make(map[string]string, len(parsed.Entries))
		for _, entry := range parsed.Entries {
			got[entry.Key] = entry.Value
		}

		for name, want := range hostile {
			if got[name] != want {
				t.Errorf("%s: %s\n got %q\nwant %q", target.format, name, got[name], want)
			}
		}
		if len(got) != len(hostile) {
			t.Errorf("%s: round trip produced %d values, want %d", target.format, len(got), len(hostile))
		}
	}
}

// TestDockerRoundTripsWhatItCanRepresent.
//
// Docker's `--env-file` grammar has no quoting at all — everything after the
// first `=` is the value, verbatim, to the end of the line — so it is parsed
// back with the dotenv parser only for values that contain no line break and no
// leading or trailing whitespace. The formats differ, and pretending otherwise
// is how a value comes back changed.
func TestDockerRoundTripsWhatItCanRepresent(t *testing.T) {
	representable := map[string]string{
		"SIMPLE":    "hunter2",
		"EQUALS":    "a=b=c",
		"DOLLAR":    "cost $5",
		"NON_ASCII": "café",
	}

	document, err := formatSecrets(representable, "docker")
	if err != nil {
		t.Fatal(err)
	}

	for name, want := range representable {
		line := name + "=" + want
		if !strings.Contains(string(document), line+"\n") {
			t.Errorf("docker output does not carry %q", line)
		}
	}
}

// TestFormatsRefuseWhatTheyCannotRepresent.
//
// Emitting a value silently cut at the first newline is exactly the failure this
// file exists to prevent, so it is an error — and the message names the format
// that *can* carry it, because "cannot export" with no way forward is not an
// answer.
func TestFormatsRefuseWhatTheyCannotRepresent(t *testing.T) {
	for _, refusal := range []struct {
		format string
		value  string
		reason string
	}{
		{"shell", carriageReturn, "single quotes have no escape mechanism, so a CR would be written raw and lost on the way back in"},
		{"docker", "two\nlines", "Docker would read the remainder as a new variable"},
		{"docker", carriageReturn, "same"},
	} {
		_, err := formatSecrets(map[string]string{"VALUE": refusal.value}, refusal.format)
		if err == nil {
			t.Errorf("%s accepted a value it cannot represent (%s)", refusal.format, refusal.reason)
			continue
		}
		if !strings.Contains(err.Error(), "VALUE") {
			t.Errorf("%s: the refusal does not name the secret: %v", refusal.format, err)
		}
	}

	// `env`, `json` and `yaml` escape it and are the formats to reach for.
	for _, format := range []string{"env", "json", "yaml"} {
		document, err := formatSecrets(map[string]string{"VALUE": carriageReturn}, format)
		if err != nil {
			t.Errorf("%s refused a carriage return it can escape: %v", format, err)
			continue
		}
		if strings.Contains(string(document), "\r") {
			t.Errorf("%s wrote a raw carriage return rather than escaping it", format)
		}
	}
}

// TestNamesAreOrderedAndValidated.
func TestNamesAreOrderedAndValidated(t *testing.T) {
	document, err := formatSecrets(map[string]string{"ZED": "1", "ALPHA": "2", "MID": "3"}, "env")
	if err != nil {
		t.Fatal(err)
	}
	// Sorted by code unit, never by locale, so the output does not depend on
	// where it ran — and a secret added today produces a one-line diff.
	if string(document) != "ALPHA=2\nMID=3\nZED=1\n" {
		t.Errorf("output = %q", document)
	}

	// A name that is not an identifier would produce a line that parses back as
	// something else: `A=B=value` reads as `A` holding `B=value`.
	for _, name := range []string{"has-dash", "1leading", "has space", "", "has=equals"} {
		if _, err := formatSecrets(map[string]string{name: "x"}, "env"); err == nil {
			t.Errorf("accepted %q as a secret name", name)
		}
	}
}

// TestJsonMatchesJavaScriptsRendering.
//
// Go escapes `<`, `>` and `&` by default and JavaScript does not. Both parse
// back to the same string, so this is not a correctness bug — it is a
// gratuitous difference between the two modes in a file people read, and the
// kind of thing that makes somebody wonder which one is wrong.
func TestJsonMatchesJavaScriptsRendering(t *testing.T) {
	document, err := formatSecrets(map[string]string{"HTML": "<a href=\"x\">&amp;</a>"}, "json")
	if err != nil {
		t.Fatal(err)
	}
	if strings.Contains(string(document), `\u003c`) {
		t.Errorf("HTML escaping leaked into the JSON output: %s", document)
	}
	if !strings.HasSuffix(string(document), "\n") {
		t.Error("the JSON document does not end in a newline")
	}
}
